### PR TITLE
Add `pixi self-update`

### PR DESCRIPTION
**Pull request by hadim**
Supersede #582 

---

You can test it locally in your dev env with:

```bash
$ pixi run cargo run -- self-update --force --version 0.8.0
$ ./target/debug/pixi --version
pixi 0.8.0
```

---

See the current API:

```bash
$ pixi self-update
✔ Pixi will be updated from '0.1.0' to '0.12.0'
✔ Pixi archive downloaded.
✔ Pixi archive uncompressed.
✔ Pixi has been updated to version 0.12.0.
```

```bash
$ pixi self-update # wrong pixi location
✘ pixi is not installed in the default location:
- Default pixi location: /Users/hadim/.pixi/bin/pixi
- Pixi location detected: /Users/hadim/Code/libs/pixi/pixi/target/debug/pixi

It can happen when pixi has been installed via a dedicated package manager (such as Homebrew on macOS).
You can always use `pixi self-update --force` to force the update.
```

```bash
$ pixi self-update --version 0.8.0
✔ Pixi will be updated from '0.1.0' to '0.8.0'
✔ Pixi archive downloaded.
✔ Pixi archive uncompressed.
✔ Pixi has been updated to version 0.8.0.
```

```bash
$ pixi self-update
✔ pixi is already up-to-date (version 0.12.0)
```

```bash
$ pixi self-update --version 0.8.99
✘ The version you specified is not available: 0.8.99
```

### CODE REVIEW

File: src/cli/self_update.rs
Comment (wolfv) [inline src/cli/self_update.rs:264]: not sure if this needs to be `pixi.exe` on Windows. I can check

File: src/cli/self_update.rs
Comment (hadim) [inline src/cli/self_update.rs:73]: I suggest we go with that logic for now as it will work 99% of the time for the downstream users, and it's always possible to adapt the logic in the future if you get user's reports about edge cases.

Meanwhile, the `--force` flag will always allow people to update.

That being said, I am fine modifying it if you prefer.

File: src/cli/self_update.rs
Comment (hadim) [inline src/cli/self_update.rs:123]: I am using `std::process::exit(1);` in a couple of places in the code.

Is that ok with you? I am still struggling to find the best error handling logic. I tried a bit with `miette` but could not make it work.

Also should the exit code be `0`? or something else?

File: src/cli/self_update.rs
Comment (hadim) [inline src/cli/self_update.rs:100]: The version comparison is basic, but I think it should be fine:

- it's simple (simpler is better xD)
- It allows downgrading the pixi version
- at this point in the code we already know the target version is valid and exists.

Let me know if you prefer something else.

File: src/cli/self_update.rs
Comment (hadim) [reply src/cli/self_update.rs:260]: Ah good point. I don't have a windows dev machine. I'll wait for your feedback.

Comment (hadim) [reply src/cli/self_update.rs:260]: Hum the check is done with `starts_with` so I guess it should be fine:

```rust
    std::env::current_exe()
        .unwrap()
        .to_str()
        .unwrap()
        .starts_with(default_binary_path.to_str().unwrap())
```

Comment (wolfv) [reply src/cli/self_update.rs:263]: yeah, it's pixi.exe. The `starts_with` on Path/ PathBuf compares the components so I don't think this would work :)

**Comment by wolfv** [comment]
I am happy with the logic. I'll let Bas and Ruben chime in with their thoughts!

File: src/cli/self_update.rs
Comment (hadim) [reply src/cli/self_update.rs]: i fixed it - should be fine now.

File: src/cli/self_update.rs
Comment (baszalmstra) [reply src/cli/self_update.rs:120]: Yeah I think it would be better to just return a miette Error instead.

File: src/cli/self_update.rs
Comment (baszalmstra) [inline src/cli/self_update.rs:254]: You can use https://doc.rust-lang.org/std/env/consts/constant.EXE_EXTENSION.html

Comment (hadim) [reply src/cli/self_update.rs:254]: good catch - will use `std::env::consts::EXE_SUFFIX`

File: src/cli/self_update.rs
Comment (hadim) [reply src/cli/self_update.rs]: I used `Err(miette::miette!(error_mesage))?` - it's indeed much better!

**Comment by hadim** [comment]
@baszalmstra ready for another round!

I removed all the `unwrap` and replaced most of them by `except` (when the error is quite unlikely to happen).

I am also now using `miette` for "expected" errors.

I also replaced all the `println` to `eprintln`.

**Comment by hadim** [comment]
> I also replaced all the println to eprintln.

I am actually unsure whether those should be printed to stdout or stderr looking at other cli commands.

**Review by baszalmstra — COMMENTED** [review]
Some small nitpicks but looks good to me!

File: src/cli/self_update.rs
Comment (hadim) [reply src/cli/self_update.rs:69]: Thanks.

File: src/cli/self_update.rs
Comment (hadim) [reply src/cli/self_update.rs:80]: Good to know.

**Comment by hadim** [comment]
Thanks @baszalmstra - should be good now.

**Review by wolfv — APPROVED** [review]

**Comment by hadim** [comment]
> @hadim maybe you can follow up with a few docs. Great feature! :)

I already added the cli doc at `docs/cli.md` in that PR. Do you mean something else?

**Comment by wolfv** [comment]
perfect, nope. sorry. didn't look closely enough 🤦 (getting late over here haha)

**Pull request merged**
